src/note/asset_base.rs

```diff
@@ -0,0 +1,230 @@
+use alloc::vec::Vec;
+use blake2b_simd::{Hash as Blake2bHash, Params};
+use core::cmp::Ordering;
+use core::hash::{Hash, Hasher};
+use group::{Curve, Group, GroupEncoding};
+use nonempty::NonEmpty;
+use pasta_curves::arithmetic::CurveAffine;
+use pasta_curves::{arithmetic::CurveExt, pallas};
+use rand_core::CryptoRngCore;
+use subtle::{Choice, ConstantTimeEq, CtOption};
+
+use crate::{
+    constants::fixed_bases::{
+        NATIVE_ASSET_BASE_V_BYTES, VALUE_COMMITMENT_PERSONALIZATION, ZSA_ASSET_BASE_PERSONALIZATION,
+    },
+    issuance::compute_asset_desc_hash,
+    issuance_auth::{IssueAuthKey, IssueValidatingKey, ZSASchnorr},
+};
+
+/// Note type identifier.
+#[derive(Clone, Copy, Debug, Eq)]
+pub struct AssetBase(pallas::Point);
+
+// AssetBase must implement PartialOrd and Ord to be used as a key in BTreeMap.
+impl PartialOrd for AssetBase {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for AssetBase {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_coord = self.0.to_affine().coordinates().unwrap();
+        let other_coord = other.0.to_affine().coordinates().unwrap();
+        self_coord
+            .x()
+            .cmp(other_coord.x())
+            .then_with(|| self_coord.y().cmp(other_coord.y()))
+    }
+}
+
+/// Personalization for the ZSA asset digest generator
+pub const ZSA_ASSET_DIGEST_PERSONALIZATION: &[u8; 16] = b"ZSA-Asset-Digest";
+
+///    AssetDigest for the ZSA asset
+///
+///    Defined in [ZIP-227: Issuance of Zcash Shielded Assets][assetdigest].
+///
+///    [assetdigest]: https://zips.z.cash/zip-0227.html#specification-asset-identifier-asset-digest-and-asset-base
```

Comment:
@str4d ([Link](https://github.com/zcash/orchard/pull/471#discussion_r2550352041)): ```suggestion
/// Derives the Asset Digest for the given ZSA asset.
///
/// Defined in [ZIP-227: Issuance of Zcash Shielded Assets][assetdigest].
///
/// [assetdigest]: https://zips.z.cash/zip-0227#asset-digests
```